optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/DroolsConstraintSessionFactory.java

```diff
@@ -103,41 +92,98 @@ public boolean isDroolsAlphaNetworkCompilationEnabled() {
         return droolsAlphaNetworkCompilationEnabled;
     }
 
-    @Override
-    public KieSession buildSession(boolean constraintMatchEnabled, Solution_ workingSolution) {
-        ScoreDefinition<Score_> scoreDefinition = solutionDescriptor.getScoreDefinition();
-        AbstractScoreHolder<Score_> scoreHolder = scoreDefinition.buildScoreHolder(constraintMatchEnabled);
-        // Determine which rules to enable based on the fact that their constraints carry weight.
+    public SessionDescriptor<Score_> buildSession(boolean constraintMatchEnabled, Solution_ workingSolution) {
         Score_ zeroScore = scoreDefinition.getZeroScore();
-        Set<String> disabledConstraintIdSet = new LinkedHashSet<>(0);
-        compiledRuleToConstraintMap.forEach((compiledRule, constraint) -> {
-            Score_ constraintWeight = (Score_) constraint.extractConstraintWeight(workingSolution);
-            scoreHolder.configureConstraintWeight(compiledRule, constraintWeight);
-            if (constraintWeight.equals(zeroScore)) {
-                disabledConstraintIdSet.add(constraint.getConstraintId());
-            }
-        });
-        // Determine the KieBase to use.
-        if (disabledConstraintIdSet.isEmpty()) { // Shortcut; don't change the original KieBase.
-            currentKieBase = originalKieBase;
-            currentlyDisabledConstraintIdSet = null;
-        } else if (!disabledConstraintIdSet.equals(currentlyDisabledConstraintIdSet)) {
-            // Only rebuild the active KieBase when the set of disabled constraints changed.
-            ModelImpl model = new ModelImpl().withGlobals(originalModel.getGlobals());
-            constraintToModelRuleMap.forEach((constraintId, modelRule) -> {
-                if (disabledConstraintIdSet.contains(constraintId)) {
-                    return;
-                }
-                model.addRule(modelRule);
-            });
-            currentKieBase = buildKieBaseFromModel(model, droolsAlphaNetworkCompilationEnabled);
-            currentlyDisabledConstraintIdSet = disabledConstraintIdSet;
+
+        // Extract constraint weights, excluding constraints where weight is zero.
+        Map<DroolsConstraint<Solution_>, Score_> constraintWeightMap = constraintList.stream()
+                .map(constraint -> {
+                    Object weight = constraint.extractConstraintWeight(workingSolution); // Expensive, only do once.
```

Comment:
Can this ever return null? I don't see the usual "@return never null" javadoc on the `extractConstraintWeight`.

Reply:
Deep down, a validator will throw if the weight is null.